libnmstate/nm/dns.py

```diff
@@ -0,0 +1,104 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+from operator import itemgetter
+from operator import methodcaller
+
+from libnmstate import iplib
+from libnmstate.nm import nmclient
+from libnmstate.schema import DNS
+from libnmstate.nm import ipv4 as nm_ipv4
+from libnmstate.nm import ipv6 as nm_ipv6
+from libnmstate.error import NmstateInternalError
+
+
+DNS_DEFAULT_PRIORITY_VPN = 50
+DNS_DEFAULT_PRIORITY_OTHER = 100
+
+IPV6_ADDRESS_LENGTH = 128
+
+
+def get_running():
+    dns_state = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    client = nmclient.client()
+    dns_confs = client.get_dns_configuration()
+    for dns_conf in sorted(dns_confs, key=methodcaller('get_priority')):
+        iface_name = dns_conf.get_interface()
+        for ns in dns_conf.get_nameservers():
+            if iplib.is_ipv6_link_local_addr(ns, IPV6_ADDRESS_LENGTH):
+                if not iface_name:
+                    # For IPv6 link local address, the interface name should be
+                    # appended also.
+                    raise NmstateInternalError(
+                        "Got empty interface name for IPv6 link local "
+                        "DNS nameserver entry {}".format(ns))
+                ns_addr = "{}%{}".format(ns, iface_name)
+            else:
+                ns_addr = ns
+            dns_state[DNS.SERVER].append(ns_addr)
+        dns_state[DNS.SEARCH].extend(dns_conf.get_domains())
+    if not dns_state[DNS.SERVER] and dns_state[DNS.SEARCH]:
```

Comment:
Is this condition correct? Do you maybe mean `not dns_state[DNS.SERVER] and not dns_state[DNS.SEARCH]`?

Another possiblitly is to do:

```
dns_state = {}
...
if something:
     dns_state.setdefault(DNS.SERVER, []).append(ns_addr)
....
return dns_state
```

Then you do not need the extra if at the end because the dict will be empty anyway. It makes the idea more clear IMHO. However, if only server or search is specified, the other key will not appear in the dict. Is this a problem? Should it be there if it is empty?

 What do you think?

Reply:
I will just use the `if not server and not search:`